### PR TITLE
Masterbar: cleanup and optimize the "Write" button code

### PR DESCRIPTION
**Pull request by jsnajdr**
- avoid doing any wasted work on the `SitesPopover` component when the popover is not displayed
- `setPostButtonRef` sets an instance property when called, instead of updating component state -- that's not needed when setting/unsetting refs
- avoid ESlint warnings by creating `preloadPostEditor` and `closeSitesPopover` methods. That avoids creating a new function instance on each render.

**How to test:**
No behavior change intended: check that the "Write" button in masterbar works as intended and clicking on it shows the site selector. Or, in case of single-site account, opens the post editor directly.

### CODE REVIEW

File: client/layout/masterbar/publish.jsx [hunk 1]
```diff
@@ -74,25 +98,17 @@ class MasterbarItemNew extends React.Component {
...
-					<SitesPopover
-						id="popover__sites-popover-masterbar"
-						visible={ this.state.isShowingPopover }
-						context={ this.state.postButtonContext }
-						onClose={ this.toggleSitesPopover.bind( this, false ) }
-						onSiteSelect={ this.props.siteSelected }
-						groups={ true }
-						position={ this.getPopoverPosition() }
-					/>
+					{ this.renderPopover() }
```
Comment (ockham) [inline client/layout/masterbar/publish.jsx]: Late to the party and drive-by-ish, but I generally prefer `{ this.state.isShowingPopover && <SitesPopover /*...*/ />` over `this.renderSomething()` methods -- the latter just tend to accumulate and make things more abstract to read IMO.

Comment (jsnajdr) [reply client/layout/masterbar/publish.jsx]: Yes, I agree that it can be hard to read the source of a class that has many `render*` methods. The best way how to structure the code and avoid too many `render*` methods is probably extracting a separate functional component?

In this case, my motivation was to separate the "`MasterbarItem` object with many props" from the "`SitePopover` component with many props" and not have too much nesting in the main `render` method.
